### PR TITLE
Refactor: Extract parsing of document and definition nodes into separate functions

### PR DESCRIPTION
**Pull request by darrellwarde**
# Description

Slowly moving functionality out of `makeAugmentedSchema`, this PR moves the early parsing of the type definitions into a document, and the subsequent parsing of that document into GraphQL definition nodes into separate functions.

Additionally, likely improve performance by moving several `filter` calls into one `reduce`.

### CODE REVIEW

File: packages/graphql/src/schema/make-augmented-schema.ts [hunk 1]
```diff
@@ -78,14 +67,21 @@ import getUniqueFields from "./get-unique-fields";
...
+import parseFulltextDirective from "./parse/parse-fulltext-directive";
 import { QueryOptionsDirective } from "../classes/QueryOptionsDirective";
+import { parseQueryOptionsDirective } from "./parse/parse-query-options-directive";
 
 function makeAugmentedSchema(
     typeDefs: TypeSource,
     { enableRegex, skipValidateTypeDefs }: { enableRegex?: boolean; skipValidateTypeDefs?: boolean } = {}
 ): { nodes: Node[]; relationships: Relationship[]; typeDefs: DocumentNode; resolvers: IResolvers } {
-    const document = mergeTypeDefs(Array.isArray(typeDefs) ? (typeDefs as string[]) : [typeDefs as string]);
+    const document = getDocument(typeDefs);
```
Comment (angrykoala) [inline packages/graphql/src/schema/make-augmented-schema.ts:84]: :clap:

File: packages/graphql/src/schema/get-definition-nodes.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+export function getDefinitionNodes(document: DocumentNode): DefinitionNodes {
+    return document.definitions.reduce<DefinitionNodes>(
+        (definitionNodes, definition) => {
+            switch (definition.kind) {
+                case Kind.SCALAR_TYPE_DEFINITION:
+                    definitionNodes.scalarTypes.push(definition);
+                    break;
+                case Kind.OBJECT_TYPE_DEFINITION:
+                    if (!["Query", "Mutation", "Subscription"].includes(definition.name.value)) {
```
Comment (angrykoala) [inline packages/graphql/src/schema/get-definition-nodes.ts:54]: **nitpick**: Maybe `["Query", "Mutation", "Subscription"]` could be a constant, as it is used in some other places, but mainly to give a context on why these are not pushed to objectTypes, something like:
```suggestion
                    if (!nativeTopLevelTypes.includes(definition.name.value)) {
```
-----

**suggestion(non-blocking)**: Furthermore, a small function could make this really legible (you probably can think of better naming):

```typescript
function isNativeRootType(type: DefinitionNode): boolean
```

```suggestion
                    if (!isNativeRootType(definition)) {
```

File: packages/graphql/src/schema/make-augmented-schema.ts [hunk 3]
```diff
@@ -168,37 +164,21 @@ function makeAugmentedSchema(
 
     const customResolvers = getCustomResolvers(document);
 
-    const scalars = document.definitions.filter((x) => x.kind === "ScalarTypeDefinition") as ScalarTypeDefinitionNode[];
-
-    const objectNodes = document.definitions.filter(
-        (x) => x.kind === "ObjectTypeDefinition" && !["Query", "Mutation", "Subscription"].includes(x.name.value)
-    ) as ObjectTypeDefinitionNode[];
+    const definitionNodes = getDefinitionNodes(document);
```
Comment (angrykoala) [inline packages/graphql/src/schema/make-augmented-schema.ts:167]: Another :clap:  for you

**Review by angrykoala — APPROVED** [review]
Approved, just a nitpick that I would like to get solved so it is :100:

File: packages/graphql/src/schema/get-definition-nodes.ts (hunk 2 above)
Comment (darrellwarde) [reply packages/graphql/src/schema/get-definition-nodes.ts]: This is now done! For now, I've left it as a util function within the file as this is the only place where it is used. We can extract it out to a utils directory if and when it gets used elsewhere.

**Pull request merged**